internal/kubernetes/discovery.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Crunchy Data Solutions, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+
+	"github.com/crunchydata/postgres-operator/internal/logging"
+)
+
+type Version = version.Info
+
+// DiscoveryRunner implements [APIs] by reading from a Kubernetes API client.
+// Its methods are safe to call concurrently.
+type DiscoveryRunner struct {
+	Client interface {
+		ServerGroups() (*metav1.APIGroupList, error)
+		ServerResourcesForGroupVersion(string) (*metav1.APIResourceList, error)
+		ServerVersion() (*version.Info, error)
+	}
+
+	refresh time.Duration
+
+	want []API
+	have struct {
+		sync.RWMutex
+		APISet
+		Version
+	}
+}
+
+// NewDiscoveryRunner creates a [DiscoveryRunner] that periodically reads from
+// the Kubernetes at config.
+func NewDiscoveryRunner(config *rest.Config) (*DiscoveryRunner, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(config)
+
+	runner := &DiscoveryRunner{
+		Client:  dc,
+		refresh: 10 * time.Minute,
```

Comment:
Have you done a manual test (more for fun, really) where you log what the runner has, add some CRD, and see it log the new APISet after the refresh time?

Reply:
👍🏻 I hadn't, and now I have. I added the VolumeSnapshots CRDs; they appeared. I removed the CRDs; they disappeared.

⛏️ This should log when it's successful. My first attempt to dump the set using [UnsortedList](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set.UnsortedList) looked terrible.
🌱 I'll do something small and leave a TODO.

Reply:
I've added a debug log message.